src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -15,14 +15,38 @@
 #include "ray/core_worker/transport/direct_task_transport.h"
 
 #include "ray/core_worker/transport/dependency_resolver.h"
-#include "ray/core_worker/transport/direct_actor_transport.h"
 
 namespace ray {
 
 Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
   RAY_LOG(DEBUG) << "Submit task " << task_spec.TaskId();
   resolver_.ResolveDependencies(task_spec, [this, task_spec]() {
     RAY_LOG(DEBUG) << "Task dependencies resolved " << task_spec.TaskId();
+    if (RayConfig::instance().gcs_service_enabled() && task_spec.IsActorCreationTask()) {
+      RAY_CHECK(actor_create_callback_ != nullptr);
+      // If gcs actor management is enabled, the actor creation task will be sent to gcs
+      // server directly after the in-memory dependent objects are resolved.
+      // For more details please see the protocol of actor management based on gcs.
+      // https://docs.google.com/document/d/1EAWide-jy05akJp6OMtDn58XOK7bUyruWMia4E-fV28/edit?usp=sharing
+      auto actor_id = task_spec.ActorCreationId();
+      RAY_LOG(INFO) << "Submitting actor creation task to GCS: " << actor_id;
+      auto status = actor_create_callback_(task_spec, [actor_id](Status status) {
+        if (status.ok()) {
+          RAY_LOG(INFO) << "Actor creation task submitted to GCS: " << actor_id;
+        } else {
+          // If GCS is failed, GcsRpcClient may receive IOError status but it will not
+          // trigger this callback, because GcsRpcClient has retry logic at the
+          // bottom. So it must be something wrong with the protocol of gcs-based
+          // actor management if status is not OK.
+          // Just log something and exit.
+          RAY_LOG(FATAL) << "Failed to create actor " << actor_id << " with status "
+                         << status;
+        }
+      });
+      RAY_CHECK_OK(status);
+      return;
```

Comment:
> I just remembered, we should probably also call back to the TaskManager here, as we do for the non-GCS service path. This is used to indicate that the actor creation task has finished. We can also call the failure handler if the status returned an error, instead of crashing here.
> 
> There is a good example of what to do [here](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/transport/direct_task_transport.cc#L203).

GcsRpcClient is a reliable client, it will retry at the bottom and ensure that the message must be delivered to the GCS Server, even if the GCS Server is restarted. So I assert here that status must be OK, if not then there must be a serious bug in the system.

Reply:
I understand, I meant that we need to notify the TaskManager that the actor creation task is done. Otherwise, the TaskManager will still think that the creation task is pending. You have to call something like `task_finisher_->CompletePendingTask` (see the linked code for an example).

Reply:
> I understand, I meant that we need to notify the TaskManager that the actor creation task is done. Otherwise, the TaskManager will still think that the creation task is pending. You have to call something like `task_finisher_->CompletePendingTask` (see the linked code for an example).

Okey, I know what you mean, I will fix it.